feat(dev-infra): validate deprecation notes in commit messages

devversion opened the pull request:
Currently the commit message validation tool from `ng-dev` validates
the `BREAKING CHANGE:` commit message notes. This commit adds a similar
check for `DEPRECATED:` commit message notes.

Additionally, the check for breaking changes is reworked slightly to
be more tolerant (i.e. if there is only a single line break after the
summary; this is acceptable as per the parser and commonly done in the
COMP repo). The checks have been updated to capture wrong keywords that
are commonly used instead of the correct one. e.g. if a commit message
uses `DEPRECATIONS:` instead of `DEPRECATED:`, the validation will fail.

This prevents changelog generation issues where breaking change notes,
or deprecations are missing. This happened in the COMP repo where
the `DEPRECATED:` keyword was used incorrectly. See:

https://github.com/angular/components/commit/99391e79391d20c6ef2f95a3ea4fd6901dcb631d

devversion committed:
feat(dev-infra): validate deprecation notes in commit messages

Currently the commit message validation tool from `ng-dev` validates
the `BREAKING CHANGE:` commit message notes. This commit adds a similar
check for `DEPRECATED:` commit message notes.

Additionally, the check for breaking changes is reworked slightly to
be more tolerant (i.e. if there is only a single line break after the
summary; this is acceptable as per the parser and commonly done in the
COMP repo). The checks have been updated to capture wrong keywords that
are commonly used instead of the correct one. e.g. if a commit message
uses `DEPRECATIONS:` instead of `DEPRECATED:`, the validation will fail.

This prevents changelog generation issues where breaking change notes,
or deprecations are missing. This happened in the COMP repo where
the `DEPRECATED:` keyword was used incorrectly. See:

https://github.com/angular/components/commit/99391e79391d20c6ef2f95a3ea4fd6901dcb631d

gkalpak commented on dev-infra/commit-message/validate.ts:
```diff
@@ -26,16 +26,29 @@ export interface ValidateCommitMessageResult {
+ *   - `BREAKING-CHANGES: <some-content>` | The wrong keyword is used here.
+ */
+const INCORRECT_BREAKING_CHANGE_BODY_RE =
+    /^(BREAKING CHANGE[^:]|BREAKING-CHANGE|BREAKING[ -]CHANGES)/m;
```
[Nit] This could shortened as:

```suggestion
    /^(BREAKING CHANGE[^:]|BREAKING-CHANGES?)/m;
```

devversion replied:
Yeah, good point. I've played with simplifying this before but felt like being explicit is better here. Let me know if you feel strongly about changing this. Happy to change.

josephperrott reviewed (approved):
LGTM

I feel like there is a more generic way to do this (more towards having "approved" keywords but detecting all keywords, but I am not sure of how to do it, nor if its worth the effort required.

devversion commented:
Agreed. This could have been made more generic, but if we do this (in a way where we detect the format of: `<keyword>: <msg>`) then we might also accidentally report an invalid commit message if someone uses a commit message with bullet points (e.g `changes being made:`)

AndrewKushnir commented:
@devversion @josephperrott  FYI I've tried merging this PR and got the following error:

```
Could not merge the specified pull request.
Pull Request has at least one commit containing a breaking change note, but does not have a breaking change label.
```

It looks like the `BREAKING CHANGES` appears in the commit message, thus the merge script fails.
**Please let me know if I should just add the "breaking changes" label and merge.**

The only reason I decided to hold on is to double check whether this would somehow trigger the commit message to be included into the log (with the breaking changes note). Thus this is unlikely since it's a dev-infra commit (which we filter out).

devversion committed:
fix(dev-infra): breaking change and deprecation notes incorrectly picked up

If a commit message currently mentions the breaking change or
deprecation note keywords, the commit message parse logic
accidentally picks up the note. This could then accidentally
prevent the commit from being merged (e.g. if the commit targets
the patch branch but mentioned the `BREAKING CHANGE: ` marker).

This commit switches the commit message notes pattern to only
capture notes at the beginning of a line (also allowing accidental
whitespace). This matches with the format we describe in our
contribution guide, as well as with our commit message validation
logic that also assumes notes at the beginning of a line.

devversion commented:
@AndrewKushnir Great observation that forcibly proceeding with merge would have caused this commit to show up in the changelog. I've added a second commit that fixes the commit-message parsing logic, so that we can land this change.

@AndrewKushnir @josephperrott Please have another look.

josephperrott reviewed (approved):
LGTM
